### PR TITLE
add azure workload identity support

### LINKED ISSUE #3698: Add support for backup on Microsoft Azure Blob Storage using Workload Identity
Output of `restic version`
--------------------------
`0.13.0`

What should restic do differently? Which functionality do you think we should add?
----------------------------------------------------------------------------------
In[ Azure AD Workload Identity](https://azure.github.io/azure-workload-identity/docs/) model the Kubernetes cluster becomes a token issuer, issuing tokens to Kubernetes Service Accounts. These service account tokens can be configured to be trusted on Azure AD applications. So Restic shouldn't require the storage access credentials and use the token instead when it doesn't have the credentials. 

What are you trying to do? What problem would this solve?
---------------------------------------------------------
I'm trying to backup some data with restic from a workload deployed in AKS cluster using workload identity federation. But the backup is failing because Restic requires access credentials to store backups on Microsoft Azure Blob Storage. 

Did restic help you today? Did it make you happy in any way?
------------------------------------------------------------
Yes. It's helping us a lot.

### PR DESCRIPTION
**Pull request by ekarlso**
- Add support for Azure Workload identy - fixes #3698

What does this PR change? What problem does it solve?
-----------------------------------------------------

Adds support for workload identity in Azure AKS which is replacing Pod Identity and gives you password-less authentication towards Azure using AzureAD Credential Federation

Was the change previously discussed in an issue or on the forum?
----------------------------------------------------------------

Closes #3698

Checklist
---------

- [x] I have read the [contribution guidelines](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#providing-patches).
- [X] I have [enabled maintainer edits](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork).
- [x] I have added tests for all code changes.
- [x] I have added documentation for relevant changes (in the manual).
- [x] There's a new file in `changelog/unreleased/` that describes the changes for our users (see [template](https://github.com/restic/restic/blob/master/changelog/TEMPLATE)).
- [X ] I have run `gofmt` on the code in all commits.
- [x] All commit messages are formatted in the same style as [the other commits in the repo](https://github.com/restic/restic/blob/master/CONTRIBUTING.md#git-commits).
- [x] I'm done! This pull request is ready for review.

### CODE REVIEW

File: internal/backend/azure/config.go [hunk 1]
```diff
@@ -51,5 +57,11 @@ func ParseConfig(s string) (interface{}, error) {
 	cfg := NewConfig()
 	cfg.Container = container
 	cfg.Prefix = prefix
+
+	cfg.ClientID = os.Getenv("AZURE_CLIENT_ID")
```
Comment (MichaelEischer) [inline internal/backend/azure/config.go:61]: Setting config fields based on environment variable belongs at https://github.com/restic/restic/blob/cccc17e4e91facdaf710dc82a59653c5c0de3156/cmd/restic/global.go#L622 .

Comment (MichaelEischer) [inline internal/backend/azure/config.go:61]: Also please document the variables at https://github.com/restic/restic/blob/master/doc/040_backup.rst#environment-variables

File: internal/backend/azure/token_credential.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+// clientAssertionCredential authenticates an application with assertions provided by a callback function.
+type clientAssertionCredential struct {
+	assertion, file string
+	client          confidential.Client
+	lastRead        time.Time
+}
+
+// clientAssertionCredentialOptions contains optional parameters for ClientAssertionCredential.
+type clientAssertionCredentialOptions struct {
+	azcore.ClientOptions
```
Comment (MichaelEischer) [inline internal/backend/azure/token_credential.go:23]: With just a single field, the struct is not really necessary.

File: internal/backend/azure/token_credential.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+}
+
+// newClientAssertionCredential constructs a clientAssertionCredential. Pass nil for options to accept defaults.
+func newClientAssertionCredential(tenantID, clientID, authorityHost, file string, options *clientAssertionCredentialOptions) (*clientAssertionCredential, error) {
```
Comment (MichaelEischer) [inline internal/backend/azure/token_credential.go:27]: Is there some documentation which describes the necessary ceremony to get the credentials? Just from looking at the code, I can't tell whether it makes sense or not.

**Review by MichaelEischer — CHANGES_REQUESTED** [review]

File: internal/backend/azure/token_credential.go (hunk 2 above)
Comment (ekarlso) [reply internal/backend/azure/token_credential.go:23]: Yeah, removed it and added clientOpts instead

File: internal/backend/azure/token_credential.go (hunk 3 above)
Comment (ekarlso) [reply internal/backend/azure/token_credential.go:27]: Sorry not except the microsoft example here: v

**Comment by ekarlso** [comment]
Any idea on this?

**Comment by ekarlso** [comment]
@MichaelEischer ? ^

File: internal/backend/azure/token_credential.go (hunk 3 above)
Comment (MichaelEischer) [reply internal/backend/azure/token_credential.go]: @ekarlso I do NOT appreciate copying large parts of an example without proper attribution as required by the license.

**Comment by MichaelEischer** [comment]
I've rewritten the PR to use [azidentity](https://github.com/Azure/azure-sdk-for-go/tree/sdk/azidentity/v1.3.0) which handles the whole Azure AD authentication stuff. This also adds several other authentication methods for free, see https://github.com/Azure/azure-sdk-for-go/tree/sdk/azidentity/v1.3.0/sdk/azidentity#defaultazurecredential .

@ekarlso Please test whether this works as expected.

**Review by MichaelEischer — APPROVED** [review]
LGTM.

I've tested with a service principal, but I'd expect Managed / Workload Identity to also work.

**Pull request merged**